src/position.rs

```diff
@@ -0,0 +1,144 @@
+use std::path::Path;
+use std::cmp;
+use std::hash::Hasher;
+use std::collections::hash_map::DefaultHasher;
+use std::fs;
+use std::str;
+use rust_htslib::bam;
+use rust_htslib::bam::{record::Aux, Read};
+use log::{trace, info};
+
+pub fn depositionify_bam(input_path: &str, output_path: &str, max_mem: u64, nthreads: usize) {
+    let mut bam = threaded_bam_reader(input_path, nthreads);
+    let header = bam::Header::from_template(bam.header());
+    let bam_bytes = fs::metadata(&input_path).unwrap().len();
+    let buckets = ((bam_bytes / max_mem) + 1) as u32;
+
+    let output = if input_path == output_path {
+        let mut hasher = DefaultHasher::new();
+        hasher.write(output_path.as_bytes());
+        format!("{}.{}.tmp", output_path, hasher.finish())
+    } else {
+        output_path.to_string()
+    };
+    
+    info! {"Processing file {} of size {} into {} buckets", input_path, bam_bytes, buckets };    
+    clean_buckets(); 
+    bucketify_bam(&mut bam, &header, buckets);
+    process_buckets(&header, &output, buckets, nthreads);
+    clean_buckets(); 
+
+    if output != output_path {
+        fs::remove_file(output_path).unwrap();
+        fs::rename(output, output_path).unwrap();
+    }
+}
+
+fn bucketify_bam(bam: &mut bam::Reader, header: &bam::Header, nbuckets: u32) {
+    fs::create_dir_all(Path::new("buckets")).unwrap();
+    let mut writers: Vec<_> = bucket_names("buckets", nbuckets).iter()
+        .map(|p| bam::Writer::from_path(p, &header, bam::Format::Bam).unwrap())
+        .collect();
+    
+    for r in bam.records() {
+        let record = match r {
+            Ok(r) => {
+                trace!("Processing record: {:?}", r);
+                r
+            }
+            Err(e) => {
+                info! { "failed to read BAM record {:?}", e };
+                continue
+            }
+        };
+        //Grab the read name and figure out which bucket it goes into
+        let qname = record.qname();
+        let mut hasher = DefaultHasher::new();
+        hasher.write(qname);
+        let hash = hasher.finish();
+        let bucket = (hash % (nbuckets as u64)) as usize;
+        //XXX: Not sure if this is buffered I/O or not, if so, we'll have to manually buffer
```

Comment:
Can this buffering question be checked with `strace`? Non-buffered I/O could have a significant impact on performance.

Reply:
We can also check the code, everything should be open source? I think it is buffered, though: if it supports threaded writes, how else would they implement that without a buffer?